signer/xpi/xpi.go

```diff
@@ -143,54 +148,116 @@ func (s *PKCS7Signer) Config() signer.Configuration {
 	}
 }
 
-// SignFile takes an unsigned zipped XPI file and returned a signed XPI file
-func (s *PKCS7Signer) SignFile(input []byte, options interface{}) (signer.SignedFile, error) {
+// SignFile takes an unsigned zipped XPI file and returns a signed XPI file
+func (s *XPISigner) SignFile(input []byte, options interface{}) (signedFile signer.SignedFile, err error) {
 	var (
-		signedFile []byte
+		pkcs7Manifest []byte
+		manifest      []byte
+		metas         = []Metafile{}
+		opt           Options
+		coseSig       []byte
+		coseSigAlgs   []*cose.Algorithm
 	)
-	manifest, sigfile, err := makeJARManifests(input)
+
+	opt, err = GetOptions(options)
+	if err != nil {
+		return nil, errors.Wrap(err, "xpi: cannot get options")
+	}
+	cn, err := opt.CN(s)
+	if err != nil {
+		return nil, err
+	}
+	coseSigAlgs, err = opt.Algorithms()
+	if err != nil {
+		return nil, errors.Wrap(err, "xpi: error parsing COSE Algorithms from options")
+	}
+
+	manifest, err = makeJARManifest(input)
+	if err != nil {
+		return nil, errors.Wrap(err, "xpi: cannot make JAR manifest from XPI")
+	}
+
+	// don't issue COSE signatures if we cannot parse valid and
+	// supported COSE Algorithms from signing request options
+	if len(coseSigAlgs) < 1 {
+		pkcs7Manifest = manifest
```

Comment:
We do error when COSE Algorithm parsing and validation fails here: https://github.com/mozilla-services/autograph/pull/76/files/a9033a46e2e67b2cdc50bfa23a3786454da48f01#diff-0bf2ee5528414c7d2ac0e24b14a748daR170

This is handling the case where no COSE Algorithms are provided and we only issue a PKCS7 signature. Since we don't need to add entries for cose.sig and cose.manifest to the PKCS7 manifest file, we pass through the manifest from the ZIP to avoid recalculating it.

I'll update the comment to make that clearer.